plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -239,10 +242,25 @@ private Content toContent(final StorageTx tx, final MavenPath mavenPath) throws
   public Content put(final MavenPath path, final Payload payload)
       throws IOException, InvalidContentException
   {
-    try (StorageTx tx = storageFacet.openTx()) {
-      Content content = put(tx, path, payload);
-      tx.commit();
-      return content;
+    try (final TempStreamSupplier streamSupplier = new TempStreamSupplier(payload.openInputStream())) {
+      return storageFacet.perform(new Operation<Content>()
+      {
+        @Override
+        public Content execute(final StorageTx tx) {
+          try {
+            return put(tx, path, payload, streamSupplier.get());
+          }
+          catch (IOException e) {
+            throw Throwables.propagate(e);
+          }
```

Comment:
this smelly very badly of api that is lacking proper throwable declarations.

wrapping exceptions is not cheap